pallets/author-filter/src/lib.rs

```diff
@@ -0,0 +1,188 @@
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Small pallet responsible determining which accounts are eligible to author at the current
+//! block height.
+//!
+//! Currently this pallet is tightly coupled to our stake pallet, but this design
+//! should be generalized in the future.
+//!
+//! Using a randomness beacon supplied by the `Randomness` trait, this pallet takes the set of
+//! currently staked accounts from pallet stake, and filters them down to a pseudorandom subset.
+//! The current technique gives no preference to any particular author. In the future, we could
+//! disfavor authors who are authoring a disproportionate amount of the time in an attempt to
+//! "even the playing field".
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::pallet;
+
+pub use pallet::*;
+
+#[pallet]
+pub mod pallet {
+
+	use frame_support::debug;
+	use frame_support::pallet_prelude::*;
+	use frame_support::traits::Randomness;
+	use frame_support::traits::Vec;
+	use frame_system::pallet_prelude::*;
+	use sp_core::H256;
+	use sp_runtime::Percent;
+
+	/// The Author Filter pallet
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// Deterministic on-chain pseudo-randomness used to do the filtering
+		type RandomnessSource: Randomness<H256>;
+		/// A source for the complete set of potential authors.
+		/// The starting point of the filtering.
+		type PotentialAuthors: Get<Vec<Self::AccountId>>;
+	}
+
+	// This code will be called by the author-inherent pallet to check whether the reported author
+	// of this block is eligible at this height. We calculate that result on demand and do not
+	// record it instorage (although we do emit a debugging event for now).
+	impl<T: Config> author_inherent::CanAuthor<T::AccountId> for Pallet<T> {
+		fn can_author(account: &T::AccountId) -> bool {
+			let mut staked: Vec<T::AccountId> = T::PotentialAuthors::get();
+
+			let num_eligible = EligibleRatio::<T>::get().mul_ceil(staked.len());
+			let mut eligible = Vec::with_capacity(num_eligible);
+
+			for i in 0..num_eligible {
+				// A context identifier for grabbing the randomness. Consists of two parts
+				// - The constant string *b"filter" - to identify this pallet
+				// - The index `i` when we're selecting the ith eligible author
+				// Currently this has the weakness that the authors are based only on para-block
+				// height. This will be aleviated in the future by adding entropy from the relay
+				// chain inherent.
+				let subject: [u8; 7] = [b'f', b'i', b'l', b't', b'e', b'r', i as u8];
+				let randomness : sp_core::H256 = T::RandomnessSource::random(&subject);
```

Comment:
This is why I want to use data from the parachain inherent when checking the author inherent.

See how we do it in Moonbeam where we use the relay block height. The problem is this logic can't be used until block execution. https://github.com/PureStake/moonbeam/blob/2427474e054f5dea16bab141622aa72ded492190/pallets/author-filter/src/lib.rs#L82-L97